internal/internal_workflow.go

```diff
@@ -444,32 +456,42 @@ func newWorkflowContext(env WorkflowEnvironment, interceptors WorkflowIntercepto
 	return rootCtx
 }
 
-func newWorkflowInterceptors(env WorkflowEnvironment, factories []WorkflowInterceptorFactory) (WorkflowInterceptor, *workflowEnvironmentInterceptor) {
+func newWorkflowInterceptors(env WorkflowEnvironment, interceptors []WorkflowInterceptor) (*workflowEnvironmentInterceptor, error) {
 	envInterceptor := &workflowEnvironmentInterceptor{env: env}
-	var interceptor WorkflowInterceptor = envInterceptor
-	for i := len(factories) - 1; i >= 0; i-- {
-		interceptor = factories[i].NewInterceptor(env.WorkflowInfo(), interceptor)
+	var interceptor WorkflowInboundCallsInterceptor = envInterceptor
+	for i := len(interceptors) - 1; i >= 0; i-- {
+		interceptor = interceptors[i].InterceptWorkflow(env.WorkflowInfo(), interceptor)
```

Comment:
This feels weird.  Why we would intercept workflow invocation when we are setting up interceptor chain?

Reply:
Because we create a chain per workflow invocation. This is pretty similar to gRPC interceptor: https://grpc.github.io/grpc-java/javadoc/io/grpc/ClientInterceptor.html